### PR TITLE
Fixed rate functions and lag ratios

### PR DESCRIPTION
**Pull request by bill-j-shillito**
## Changelog / Overview

- Fixed animations so that certain rate functions (`running_start`, `wiggle`, `ease_in_back`, `ease_out_back`, `ease_in_out_back`, `ease_in_elastic`, `ease_out_elastic`, and `ease_out_elastic`) can go outside the range from 0 to 1.
- Fixed lag ratios so that they're spaced out evenly within the time interval and the rate functions are applied to each animation individually, rather than having the rate function determine when the animation starts.
- Fixed faulty code for `ease_in_out_expo`, `ease_in_bounce`, `ease_out_bounce`, and `ease_in_out_bounce`.

## Motivation

These changes make rate functions and lag ratios work they way they seem to have been intended

## Explanation for Changes

Allowing rate functions to leave the interval [0,1] gives an object some "elasticity" for nicer-looking animations, allowing them to overshoot their target and then come back. The change to lag ratios makes them work more in line with how other software works, such as After Effects. (Before they had been described as somewhat "deceptive".) Before, the animations would all be spaced by different intervals and proceed at different speeds, following the rate function _overall_; now, the animations are spaced at even intervals and proceed at the same speed, each one moving _individually_ according to the rate function.

[Documentation Reference](https://manimce--####.org.readthedocs.build)

## Testing Status

- I made sure the [LagRatios](https://docs.manim.community/en/v0.7.0/reference/manim.animation.animation.Animation.html?highlight=lag_ratio#lagratios) example still worked.

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [ ] My new functions/classes either have a docstring or are private
- [ ] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [x] The PR title is descriptive enough
- [x] The PR is labeled correctly
- [x] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

File: manim/animation/animation.py [hunk 1]
```diff
@@ -318,7 +322,8 @@ def get_sub_alpha(self, alpha: float, index: int, num_submobjects: int) -> float
         full_length = (num_submobjects - 1) * lag_ratio + 1
         value = alpha * full_length
         lower = index * lag_ratio
-        return min(max((value - lower), 0), 1)
+        return self.rate_func(value - lower)
+        # return min(max((value - lower), 0), 1)
```
Comment (jsonvillanueva) [inline manim/animation/animation.py:326]: Remove dead code
```suggestion
```

File: manim/animation/animation.py [hunk 2]
```diff
@@ -275,8 +275,12 @@ def interpolate(self, alpha: float) -> None:
             The relative time to set the aniamtion to, 0 meaning the start, 1 meaning
             the end.
         """
-        alpha = min(max(alpha, 0), 1)
-        self.interpolate_mobject(self.rate_func(alpha))
+        # alpha = min(max(alpha, 0), 1)
```
Comment (jsonvillanueva) [inline manim/animation/animation.py:278]:
```suggestion
```

**Comment by bill-j-shillito** [comment]
By the way, here's a scene I made that showcases all of the easing functions. The ones that leave the interval [0,1] are in yellow.

```python
class EasingAllAtOnce(Scene):
    def construct(self):
        rfs = []
        broken = [
            rate_functions.wiggle,
            rate_functions.running_start,
            rate_functions.ease_in_back,
            rate_functions.ease_out_back,
            rate_functions.ease_in_out_back,
            rate_functions.ease_in_elastic,
            rate_functions.ease_out_elastic,
            rate_functions.ease_in_out_elastic
        ]
        for k, v in rate_functions.__dict__.items():
            if "function" in str(v):
                if(
                    not k.startswith("__")
                    and not k.startswith("sqrt")
                    and not k.startswith("bezier")
                    and not k.endswith("not_quite_there")
                    and not k.endswith("squish_rate_func")
                    and not k.endswith("sigmoid")
                    and not k.endswith("unit_interval")
                    and not k.endswith("zero")
                    and not k.endswith("wraps")
                ):
                    rate_func = v
                    line = Line(LEFT,RIGHT).set_color(BLUE_E)
                    title = (
                        Text(rate_func.__name__, weight=BOLD)
                        .scale(0.5)
                        .next_to(line, DOWN)
                    )
                    if v in broken:
                        title.set_color(YELLOW)
                    dot = Dot().move_to(line.get_left())
                    rfs.append([VGroup(line,dot,title),dot,line,rate_func])
        #print(rfs)
        x = VGroup(*[rf[0] for rf in rfs])
        x.arrange_in_grid(cols=7, buff=(0.5,1.5), flow_order="dr")
        x.height = config.frame_height
        x.width = config.frame_width
        x.move_to(ORIGIN).scale(0.95)
        self.add(x)
        self.play(FadeIn(x))
        def go(t=1):
            self.play(
                *[rf[1]
                    .animate(rate_func=rf[3], run_time=t)
                    .move_to(rf[2].get_right())
                    for rf in rfs
                ]
            )
        def back():
            self.play(
                *[rf[1]
                    .animate(rate_func=rate_functions.linear, run_time=0.1)
                    .move_to(rf[2].get_left())
                    for rf in rfs
                ]
            )
        self.wait(1)
        go(2)
        self.wait(1)
        back()
        self.wait(1)
        go(4)
        self.wait(1)
        self.play(FadeOut(x))
```

File: manim/animation/animation.py [hunk 3]
```diff
@@ -275,8 +275,12 @@ def interpolate(self, alpha: float) -> None:
...
+        self.interpolate_mobject(alpha)
+
+    @deprecated(until="v0.6.0", replacement="interpolate")
```
Comment (jsonvillanueva) [inline manim/animation/animation.py:281]: We're already on version 0.7.0 so this deprecated decorator should be sometime after the current version (e.g. v0.8.0)

**Review by jsonvillanueva — CHANGES_REQUESTED** [review]

**Comment by jsonvillanueva** [comment]
So there's actually some weird issues cropping up in the readthedocs (rtd) output for your PR even though that check builds successfully.

Here's your PR's output:
https://manimce--1678.org.readthedocs.build/en/1678/reference/manim.utils.rate_functions.html#ratefunctions1example
Here's the latest changes output:
https://docs.manim.community/en/latest/reference/manim.utils.rate_functions.html#ratefunctions1example

**Review by behackl — APPROVED** [review]
These changes seem good to me.

Thank you for your efforts!

**Review by jsonvillanueva — APPROVED** [review]

**Pull request merged**